packages/ohm-js/src/buildGrammar.js

```diff
@@ -19,16 +20,79 @@ function namespaceHas(ns, name) {
 // `tree`, which is the concrete syntax tree of a user-written grammar.
 // The grammar will be assigned into `namespace` under the name of the grammar
 // as specified in the source.
-export function buildGrammar(match, namespace, optOhmGrammarForTesting) {
+export function buildGrammar(match, namespace, options, optOhmGrammarForTesting) {
   const builder = new Builder();
   let decl;
   let currentRuleName;
   let currentRuleFormals;
   let overriding = false;
   const metaGrammar = optOhmGrammarForTesting || ohmGrammar;
+  
+  const fetchGrammarInternal = (path) => {
+    if (!validateOption(options, 'fetchGrammar', 'function'))
+    {
+      throw new Error("Missing option 'fetchGrammar' of type `function` when trying to include.");
+    }
+
+    const grammarContent = options.fetchGrammar(path);
+
+    if (typeof grammarContent !== "string")
+    {
+      throw new Error(`Expected string from 'fetchGrammar' function, but got ${typeof(grammarContent)}`);
+    }
+
+    return grammarContent.trim();
+  }
+
+  const rematchInput = (includes) => {
+    let modifiedGrammarSource = match.input;
+
+    for (let i = 0; i < includes.length; i++) {
+      const [sourceString, fileContent] = includes[i];
+      
+      // Always substitute the include even with a nothing to prevent infinite loop.
+      modifiedGrammarSource = modifiedGrammarSource.replace(sourceString, fileContent);
+    }
+
+    const newMatch = ohmGrammar.match(modifiedGrammarSource);
+
+    if (newMatch.failed()) {
+      throw errors.grammarSyntaxError(newMatch);
+    }
+
+    helpers(newMatch).visit()
+  }
 
   // A visitor that produces a Grammar instance from the CST.
   const helpers = metaGrammar.createSemantics().addOperation('visit', {
+    Document(includesNode, grammarsNode)
+    {
+      const resolvedIncludes = includesNode.visit();
+
+      // We need to rebuild the grammar match with the resolved includes substituted.
+      // Note: It's important we prevent any deeper visits in this tree as it's now pointless.
+      if (resolvedIncludes.length > 0) {
+        rematchInput(resolvedIncludes);
```

Comment:
This doesn't seem like the right way to handle this. Naively I'd expect a call to `compileAndLoad` for each include.

And modifying the original source and calling `match` again shouldn't be necessary — we've already got `namespace` in which we look up other grammars. Can we visit includes, and call `compileAndLoad` on each of the includes to the namespace?

Which makes me realize that there is an unresolved design decision here, about whether includes are also "exported" or not (i.e., whether or not they are present in the final namespace that's handed back to the user). For simplicity I think it's fine if they are for now, but maybe there is an argument that they shouldn't be.